File: app/src/main/java/com/islavikfx/spoof/utils/SELinuxBypass.java
package com.islavikfx.spoof.utils;
import android.content.Context;
import androidx.annotation.NonNull;
import com.topjohnwu.superuser.Shell;
import org.json.JSONObject;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;


public class SELinuxBypass {

    private static final String BYPASS = "SELinuxBypass";
    private static String binaryPath;

    public static void init(Context context) {
        binaryPath = context.getFilesDir().getAbsolutePath() + "/" + BYPASS;
        extractBinary(context);
        Shell.cmd("chmod 777 " + binaryPath).exec();
    }

    private static void extractBinary(Context context) {
        File binaryFile = new File(binaryPath);
        if (binaryFile.exists()) return;

        try (InputStream input = context.getAssets().open("bin/" + BYPASS);
             FileOutputStream output = new FileOutputStream(binaryFile)) {
            byte[] buffer = new byte[8192];
            int length;
            while ((length = input.read(buffer)) > 0) {
                output.write(buffer, 0, length);
            }
        } catch (Exception ignored) {
        }
    }

    public static boolean executeWithBypass(String command) {
        try {
            Shell.Result result = Shell.cmd(binaryPath + " bypass \"" + command + "\"").exec();
            String jsonStr = String.join("\n", result.getOut());
            JSONObject json = new JSONObject(jsonStr);
            return json.optBoolean("success", false);
        } catch (Exception e) {
            return false;
        }
    }

    @NonNull
    public static String getSelinuxStatus() {
        try {
            Shell.Result result = Shell.cmd(binaryPath + " status").exec();
            String jsonStr = String.join("\n", result.getOut());
            JSONObject json = new JSONObject(jsonStr);
            return json.optString("selinux_status", "Unknown");
        } catch (Exception e) {
            return "Unknown";
        }
    }

    public static boolean isEnforcing() {
        try {
            Shell.Result result = Shell.cmd(binaryPath + " status").exec();
            String jsonStr = String.join("\n", result.getOut());
            JSONObject json = new JSONObject(jsonStr);
            return json.optBoolean("enforcing", false);
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean installPolicy() {
        try {
            Shell.Result result = Shell.cmd(binaryPath + " install_policy").exec();
            String jsonStr = String.join("\n", result.getOut());
            JSONObject json = new JSONObject(jsonStr);
            return json.optBoolean("success", false);
        } catch (Exception e) {
            return false;
        }
    }

    public static Pair<Integer, String> executeCommand(String command) {
        try {
            Shell.Result result = Shell.cmd(binaryPath + " execute \"" + command + "\"").exec();
            String jsonStr = String.join("\n", result.getOut());
            JSONObject json = new JSONObject(jsonStr);
            return new Pair<>(json.optInt("exit_code", -1), json.optString("output", ""));
        } catch (Exception e) {
            return new Pair<>(-1, "");
        }
    }

    public record Pair<F, S>(F first, S second) {
    }

}

File: app/src/main/java/com/islavikfx/spoof/utils/SelinuxBypass.kt
package com.islavikfx.spoof.utils
import com.topjohnwu.superuser.Shell


object SelinuxBypass {

    @Suppress("unused")
    private fun findMappingAddress(): String? {
        val result = Shell.cmd("cat /proc/kallsyms | grep -E 'selinux_map|current_mapping' | head -1 | awk '{print $1}'").exec()
        if (result.isSuccess && result.out.isNotEmpty()) {
            return result.out[0].trim()
        }
        return null
    }

    @Suppress("unused")
    fun disableSelinuxViaMapping(): Boolean {
        return try {
            Shell.cmd("echo '2' > /sys/fs/selinux/allow_unknown 2>/dev/null").exec()
            Shell.cmd("echo 1 > /sys/fs/selinux/reload_policy 2>/dev/null").exec()
            true
        } catch (@Suppress("UNUSED_PARAMETER") e: Exception) {
            false
        }
    }

    @Suppress("unused")
    fun disableSelinuxWithMagisk(): Boolean {
        return try {
            Shell.cmd("magiskpolicy --live 'allow shell app_data_file dir write' 2>/dev/null").exec()
            Shell.cmd("magiskpolicy --live 'allow shell app_data_file file write' 2>/dev/null").exec()
            Shell.cmd("magiskpolicy --live 'allow shell app_data_file dir read' 2>/dev/null").exec()
            Shell.cmd("magiskpolicy --live 'allow shell app_data_file file read' 2>/dev/null").exec()
            Shell.cmd("magiskpolicy --live 'allow untrusted_app app_data_file dir write' 2>/dev/null").exec()
            Shell.cmd("magiskpolicy --live 'allow untrusted_app app_data_file file write' 2>/dev/null").exec()
            true
        } catch (@Suppress("UNUSED_PARAMETER") e: Exception) {
            false
        }
    }

    @Suppress("unused")
    fun runWithSelinuxContext(command: String): Boolean {
        val contexts = listOf("u:r:shell:s0",
            "u:r:init:s0",
            "u:r:system_app:s0",
            "u:r:su:s0")
        for (context in contexts) {
            try {
                val result = Shell.cmd("su --context $context -c \"$command\"").exec()
                if (result.isSuccess) return true
            } catch (@Suppress("UNUSED_PARAMETER") e: Exception) { }
        }
        return false
    }

    @Suppress("unused")
    fun disableWithSupolicy(): Boolean {
        return try {
            Shell.cmd("supolicy --live 'permissive *' 2>/dev/null").exec()
            true
        } catch (@Suppress("UNUSED_PARAMETER") e: Exception) {
            false
        }
    }

    @JvmStatic
    fun executeWithBypass(command: String): Boolean {
        val methods = listOf(
            { disableSelinuxViaMapping() },
            { disableSelinuxWithMagisk() },
            { disableWithSupolicy() },
            {
                runWithSelinuxContext(command)
                true
            }
        )
        for (method in methods) {
            try {
                if (method()) return true
            } catch (@Suppress("UNUSED_PARAMETER") e: Exception) { }
        }
        val result = Shell.cmd(command).exec()
        return result.isSuccess
    }

}